ouroboros-consensus/docs/diffusion-pipelining-design/block-diffusion-pipelining.tex

```diff
@@ -0,0 +1,554 @@
+(identified by certificate) \emph{only until} the current chain is successfully
+extended. This approach would avoid considering more than a single bad block from any
+recent issuer. The number of recent adversarial issuers (and thus the size of
+the tracking set) is bounded by the leadership schedule and the honest stake
+assumption. The certificate counters are bounded (or can be bounded) as
+described in \cref{many-valid-block-headers}.
+
+\pagebreak
+
+\section{Implementation changes}
+The implementation changes required to deliver on the approach discussed in this document
+are all to be made within the Consensus layer. The discussion in this section pertains specifically 
+to the "non-aggressive" diffusion pipelining variant, the specific novel behaviour here is, nodes now
+forward \emph{one unvalidated block}, so that multiple nodes can download and validate the latest 
+block \emph{concurrently}. To begin, let us remind ourselves how Cardano currently diffuses data:
+
+\begin{itemize}
+  \item With Ouroboros Praos, we want the block minted in a given slot to diffuse to the leader of the next slot early enough that they in-turn can extend the chain on-top of it.
+  \item With Praos, a node will adopt the chain which is most dense and whose intersection with the previous "best" known chain is less than $k$ blocks, where $k$ is currently set to 2160.
+  \item Cognisant of the above, the process continues as follows:
+  \begin{enumerate}
+    \item A node will notify interested peers that a new block is available by posting the new header to \lstinline{chainSync}, after block validation. 
+    \item After validation of the header locally, the downstream peer will request/pull the corresponding block body.
+    \item The block is validated, and if it satisfies checks resulting in a new longer/better chain, it will be appended as the head of our "new" chain.
+    \item This process repeats for each downstream peer.
+  \end{enumerate}
+\end{itemize}
+
+It's clear from the above description that without pipelining we incur the block validation overhead,
+scaling linearly with the number of hops a block travels during its diffusion across the critical path.
+Let us now consider the following implementation changes to data diffusion in the Consensus layer to achieve
+a faster block propagation and adoption time with pipelining, the goal is to provide a \emph{mechanism} by which we 
+can expedite a block's propagation prior to its validation whilst not invalidating existing protocol primitives:
+
+\begin{itemize}
+  \item The aim is to expedite data propagation where appropriate, and to do so without destructive change.
+  \begin{enumerate}
+    \item Extend the current API to pre-notify subscribed threads when a new desirable block is seen, prior to block body validation.
+    \item Implement a new mini-protocol which can function as an "expedited" version of \lstinline{chainSync}.
```

Comment:
There's an aspect of the argument (_not invalidating existing protocol primitives_) to keep this protocol separate from chain-sync that's worth an emphasis.  `chain-sync` will keep receiving headers which bodies are already validated. @nfrisby if this is true, will `chain-sync` need to be modified?  The `chain-sync` server will need to have access to the already validated chain of blocks, not a tentative one.

During bulk-sync phase we should use `chain-sync` stream of headers, rather than headers from the new protocol.  This means that the `FetchMode` (bluk-sync vs deadline mode) will need to be accessed by the decision logic, but that should be easy since this is already part of `NodeKernel`.